### PR TITLE
feat: tab support for indentation stripping

### LINKED ISSUE #7834: Tabs silently break indentation stripping
**Describe the bug**

Tabs are not supported in the indentation stripping parser. This is the right behavior in principle, but when a user does insert a tab, the behavior does not match their indentation. The right solution is to warn and let the user fix the problem. Supporting tabs is wrong, because tab width is undefined and will never be defined consistently.

Adding a warning may make `'' ''` strings annoying for users who have to write strings with tabs at the start of lines, but their solution is already fragile, so they should move to explicit tabs (`"\t"`) or put the string in a separate file.

**Steps To Reproduce**

1. Write a file with a `''` string indented with tabs or a mix of spaces and tabs.
2. Parsed string contains tab characters.

**Expected behavior**

Warn when the first tab is encountered in the indentation stripping logic. Suggest alternatives, for both cases: user wants tabs stripped, or (less likely, current behavior) user wants to keep tabs in their string.

**`nix-env --version` output**

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by magistau**
# Motivation

Many users prefer using hard tabs for indentation, but the existing indentation stripping logic only considers spaces as indentation characters and therefore doesn't strip any tab characters.

# Context

Closes #7834.

Previously, `stripIndentation` would look for the longest common space-consisting line prefix; now it finds either the longest common space-consisting or tab-consisting prefix. A tab after a space is not considered as indentation, neither is a space after a tab.

### CODE REVIEW

**Comment by tomberek** [comment]
First consideration is if this changes historical evaluation values.

**Comment by edolstra** [comment]
See #2911, where the conclusion was that we cannot make this incompatible change without some kind of language versioning mechanism.

**Comment by thecaralice** [comment]
> See #2911, where the conclusion was that we cannot make this incompatible change without some kind of language versioning mechanism.

This change is unlikely to break anything, as it only changes how tab-indented files get parsed. More precisely, for an indented string to be affected, every line of text in it must start with a tab and there must be no additional indentation, for example:
```nix
''
<tab>something
<tab>something else
<tab>something again
''
```

The following, however, is not affected by the change:
```nix
''
  <tab>something
  <tab>something else
  <tab>something again
''
```

File: src/libexpr/parser-state.hh [hunk 1]
```diff
@@ -175,6 +180,7 @@ inline Expr * ParserState::stripIndentation(const PosIdx pos,
     /* Figure out the minimum indentation.  Note that by design
        whitespace-only final lines are not taken into account.  (So
        the " " in "\n ''" is ignored, but the " " in "\n foo''" is.) */
```
Comment (fricklerhandwerk) [inline src/libexpr/parser-state.hh:182]:
```suggestion
       the " " in "\n ''" is ignored, but the " " in "\n foo''" is not.) */
```

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]
I gather from https://github.com/NixOS/nix/issues/7834 that the actual issue is subtle behavior happening *silently*, therefore the change mainly needs an (opt-out) warning or error on mixed indentation to reduce opportunities for problems rather than introducing more of them.

Also this needs a release note and documentation in the manual, as this would be a substantial change.

File: src/libexpr/parser-state.hh [hunk 2]
```diff
@@ -175,6 +180,7 @@ inline Expr * ParserState::stripIndentation(const PosIdx pos,
...
+    std::optional<IndentChar> indentChar = std::nullopt;
     bool atStartOfLine = true; /* = seen only whitespace in the current line */
     size_t minIndent = 1000000;
```
[the remark this answers is not in the document]
Comment (thecaralice) [reply src/libexpr/parser-state.hh:185]: It is out of scope I think but alright, will do

File: src/libexpr/parser-state.hh (hunk 1 above)
Comment (thecaralice) [reply src/libexpr/parser-state.hh:182]: Out of scope as well but valid

File: src/libexpr/parser-state.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (thecaralice) [reply src/libexpr/parser-state.hh:185]: @pennae do you suggest that someone with more than a million indent spaces will be surprised by the parser stripping more spaces than it used to?

File: tests/functional/lang/eval-okay-ind-string.exp [hunk 3]
```diff
@@ -1 +1 @@
...
... empty line after the start.\nBut there's no reason to do so: you could just repeat the empty\nline.\n  Similarly you can force an indentation level,\n  in this case to 2 spaces.  This works because the anti-quote\n  is significant (not whitespace).\nstart on network-interfaces\n\nstart script\n\n  rm -f /var/run/opengl-driver\n  ln -sf 123 /var/run/opengl-driver\n\n  rm -f /var/log/slim.log\n   \nend script\n\nenv SLIM_CFGFILE=abc\nenv SLIM_THEMESDIR=def\nenv FONTCONFIG_FILE=/etc/fonts/fonts.conf  \t\t\t\t# !!! cleanup\nenv XKB_BINDIR=foo/bin         \t\t\t\t# Needed for the Xkb extension.\nenv LD_LIBRARY_PATH=libX11/lib:libXext/lib:/usr/lib/          # related to xorg-sys-opengl - needed to load libglx for (AI)GLX support (for compiz)\n\nenv XORG_DRI_DRIVER_PATH=nvidiaDrivers/X11R6/lib/modules/drivers/ \n\nexec slim/bin/slim\nEscaping of ' followed by ': ''\nEscaping of $ followed by {: \${\nAnd finally to interpret \\n etc. as in a string: \n, \r, \t.\nfoo\n'bla'\nbar\ncut -d $'\\t' -f 1\nending dollar $$\nThis text uses\n\ttabs\nfor indentation\nAnd this text uses\n  spaces\nbut is indented with tabs\n  \tThis text uses\n\t  both spaces and tabs\n\t\tso nothing is stripped\n\t"
```
[the remark this answers is not in the document]
Comment (thecaralice) [reply tests/functional/lang/eval-okay-ind-string.exp:1]: Nothing I can help with here

File: src/libexpr/parser-state.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (thufschmitt) [reply src/libexpr/parser-state.hh]: Agreed with @pennae. This might look silly, but changes like that can have a huge impact, so we shouldn't change behaviour in any way.

**Comment by L-as** [comment]
Is there any good argument for why language versioning should be more than a `nix-language-version` setting along with an attribute in flakes to set it? Flakes are already experimental and don't need an RFC to be changed, and the `nix-language-version` setting could be guarded behind an experimental feature too.

File: src/libexpr/parser-state.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply src/libexpr/parser-state.hh]: This particular one seems very silly to me. How likely is it that code exists in the wild that relies on this behavior? We don't have a quantifiable measure of stability, but evaluation results of Nixpkgs stable releases could be a good approximation.

If we take such an absolute stance on compatibility as @pennae and @thufschmitt are suggesting, we can essentially give up on evolving the code. This *can* be a reasonable strategy, but is it desirable? And in any case it should be communicated clearly, otherwise a lot of people will waste their time trying to do things that will end up being rejected.

File: src/libexpr/parser-state.hh [hunk 4]
```diff
@@ -167,49 +168,63 @@ inline Formals * ParserState::validateFormals(Formals * formals, PosIdx pos, Sym
...
-                atStartOfLine = true;
-                curIndent = 0;
+        };
+        std::string_view view = *x;
+        for (auto cur: view) {
+            if (cur == '\n') {
+                state.reset();
+                continue;
             }
+            if (!state.indent(cur)) state.stop();
```
Comment (fricklerhandwerk) [inline src/libexpr/parser-state.hh:220]: I'm not sure about the latest change. All I suggested was simplifying the logic in place to make it more readable without impact on performance characteristics. Adding abstractions might increase the overhead measurably, and although I have no numbers to substantiate what I'm saying, I'd argue such refactorings should always come with benchmarks. Nixpkgs and NixOS are huge code bases that already suffer performance issues.

**Comment by fricklerhandwerk** [comment]
@L-as that discussion happened on https://github.com/NixOS/rfcs/pull/137. I suggest opening inline comments if you have questions, or, if what you want to add is more on the meta level, continue on the [Discourse thread](https://discourse.nixos.org/t/rfc-137-nix-language-versioning-call-for-shepherds/28918). Please read the RFC text first though, where we collected a wealth of arguments for and against certain approaches. I'd appreciate additions if you find that something is missing.

To summarise: Evolving the language is Hard(tm) given the notions backwards compatibility guarantees being discussed. But we have no agreed-upon stance, not even a clearly delineated group of people with authority to make such a decision. And even if we had the decisionmakers and decisions, we also have a resource and prioritisation problem to solve in order to implement those decisions.

File: src/libexpr/parser-state.hh (hunk 4 above)
Comment (thecaralice) [reply src/libexpr/parser-state.hh:220]: Well, I wouldn't call C++ my area of expertise, as I'm more proficient with Rust, but I don't see why would performance suffer in this case.

**Comment by Gergorius** [comment]
New user here.

I just ran into this problem and, looking at this discussion I am not seeing the light at the end of the tunnel.

What could be considered is to make the breaking change, but add a way to import nix files in a backwards compatible way. Something like an alternative `import` function. It could be called `legacyImport` (and `legacyScopedImport`). This function should take a nix version as the first argument to allow supporting more expression breaking bugfixes in the future and modify the scope of the imported nix file so it's `import` and `scopedImport` calls inherit the compatibility.

The sooner this is fixed the better because few things justify it bugging people for [6 more years](https://github.com/NixOS/nix/pull/2911), if not longer.
